kotlinx-coroutines-core/common/test/flow/terminal/IterateTest.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlin.test.*
+
+class IterateTest : TestBase() {
+    @Test
+    fun testIterateToList() = runTest {
+        val flow = flow {
+            emit(1)
+            emit(2)
+        }
+        val list = flow.iterate {
+            val mutableList = mutableListOf<Int>()
+            while (hasNext()) {
+                mutableList.add(next())
+            }
+            mutableList
+        }
+        assertEquals(listOf(1, 2), list)
+    }
+
+    @Test
+    fun testCancelsAfterDone() = runTest {
+        val flow = flow {
+            emit(1)
+            error("Should not be executed")
+        }
+        val result = flow.iterate { next() }
+        assertEquals(1, result)
+    }
+
+    @Test
+    fun testDoesNotRace() = runTest {
+        val flow = flow {
+            emit(1)
+            error("Should not be executed")
+        }
+        val result = flow.iterate {
+            next().also {
+                yield()
+                // not obvious if this results in a deterministic test?
```

Comment:
We have our own internal mechanism for making such tests explicitly deterministic: `expect(num)`, `expectUnreached()` and `finish(lastNum)`. 
But probably it's not worth changing the test in the prototype